### PR TITLE
[TCFMM-6]  Add TZIP-16 contract metadata 

### PR DESCRIPTION
**Pull request by rinn7e**
## Description

Problem: We need the contract to support TZIP-16 metadata,
the generation of it in the Haskell library as well as the documentation to do so.

Solution: Add CLI to generate metadata, add `metdata` field to storage,
add off-chain views, and add the docs.

## Related issue(s)

Resolves TCFMM-6

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by rinn7e** [comment]
@pasqu4le This PR is still WIP, but I have a question related to the off-chain view: 

> It looks like we'll need "getter" views surely for the storage's liquidity and price, we can also avoid oracle-related ones (fee_growth) as they are covered by on-chain views already.

1. For the `price` offchain view, do we refer to the `sqrt_price`, or we want the calculated price of each token?

2. Do we need tokens off-chain view, example: `tokenMetadataView` ?

**Comment by rinn7e** [comment]
@pasqu4le This PR is ready for review. The current generated metadata look like this:

```
{
    "homepage": "https://github.com/serokell/segmented-cfmm",
    "interfaces": [
        "TZIP-16"
    ],
    "views": [
        {
            "implementations": [
                {
                    "michelsonStorageView": {
                        "returnType": {
                            "prim": "nat"
                        },
                        "code": [
                            {
                                "prim": "CAR"
                            },
                            {
                                "prim": "CDR"
                            },
                            {
                                "prim": "CAR"
                            },
                            {
                                "prim": "CAR"
                            }
                        ]
                    }
                }
            ],
            "name": "get_liquidity",
            "pure": true,
            "description": "Get the liquidity of a token"
        }
    ],
    "authors": [
        "Serokell",
        "Arthur Breitman"
    ],
    "name": "Segmented CFMM Contract (Token X / Token Y)",
    "version": "0.0.1.0",
    "license": {
        "name": "MIT"
    },
    "description": "A constant product market making smart-contract, which allows the curve to be defined on price segments. Based on the ideas described in https://uniswap.org/whitepaper-v3.pdf."
}
```

`(Token X / Token Y)`: These token names could be modified by the user.

Currently, there is only one off-chain view which is `get_liquidity`. I plan to add more, hope to get your views on this: https://github.com/serokell/segmented-cfmm/pull/3#issuecomment-888361766

Let me know if I should added the default generated metadata as the artifact or not.

**Comment by pasqu4le** [comment]
@rinn7e regarding your questions:

1. I think we should have an off-chain view that returns directly the `sqrt_price`.
    The `price` can be calculated from it by off-chain apps and the same can be done for per-token prices. 
2. No, I think we shouldn't have token metadata as defined in [TZIP-12](https://gitlab.com/tezos/tzip/-/blob/master/proposals/tzip-12/tzip-12.md#token-metadata) because the only tokens that will be stored in the contract are NFTs of the positions, which are constantly chaging.

P.s. we may want to add more token-related off-chain views, but I'm pretty sure some changes in the contract are needed first, so we'll add these as we go along.

File: haskell/app/Main.hs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      mkCommandParser "print-metadata"
+        (PrintMetadata <$> metadataConfigParser)
+        "Print known part of TZIP-16 metadata."
+  ]
+
+programInfo :: Opt.ParserInfo CmdArgs
+programInfo = Opt.info (Opt.helper <*> versionOption <*> cmdArgsParser) $
+  mconcat
+  [ Opt.fullDesc
+  , Opt.progDesc "baseDAO-ligo-meta: Tools for the LIGO version of the contract."
```
Comment (pasqu4le) [inline haskell/app/Main.hs:51]: Needs an update

File: haskell/app/Main.hs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  ]
+  where
+    versionOption = Opt.infoOption ("baseDAO-ligo-meta-" <> showVersion version)
```
Comment (pasqu4le) [inline haskell/app/Main.hs:54]: Same here

File: haskell/src/SegCFMM/TZIP16Metadata.hs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  , vDescription = Just
+      "Get the square root of virtual price"
+  , vPure = Just True
+  , vImplementations =
+      [ VIMichelsonStorageView $
+          mkMichelsonStorageView @Storage @Natural Nothing [] $
+            unsafeCompileViewCode $ WithoutParam $ do
+              stToField #sSqrtPrice
+      ]
+  }
```
Comment (pasqu4le) [inline haskell/src/SegCFMM/TZIP16Metadata.hs:124]: nitpick: end-of-file newline missing

File: README.md [hunk 4]
```diff
@@ -18,4 +18,34 @@ exponentiations in the contract and verify that the precision obtained is suffic
...
-This code is work in progress and should not be expected to run or even compile.
\ No newline at end of file
+This code is work in progress and should not be expected to run or even compile.
+
+
+## Generating a contract metadata
+
+The contract metadata should be compliant with [TZIP-16](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-16/tzip-16.md).
+
+You can generate the known metadata, but you'll need the `stack` tool installed, see
```
Comment (pasqu4le) [inline README.md:28]: Not just here, but in other places there is `known` used, however we should probably not do that.

Aside from some token info (which can be provided) all metadata is known for this contract, so IMO we can remove this.

File: Makefile [hunk 5]
```diff
@@ -26,6 +26,13 @@ lib: all
 	$(MAKE) -C haskell build PACKAGE=segmented-cfmm \
 		STACK_DEV_OPTIONS="--fast --ghc-options -Wwarn"
 
+metadata : output = metadata.json
+metadata: lib all
+	$(MAKE) -C haskell exec PACKAGE=segmented-cfmm \
+		EXEC_ARGUMENTS="print-metadata --x-token-symbol x --x-token-name 'Token X' \
+		--y-token-symbol y --y-token-name 'Token Y'" \
+  		EXEC_OUTPUT=$(output)
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply Makefile]: Right, I've added more arguments to the `make metadata`.

File: haskell/app/Main.hs (hunk 1 above)
Comment (rinn7e) [reply haskell/app/Main.hs]: Fixed the description.

File: haskell/app/Main.hs (hunk 2 above)
Comment (rinn7e) [reply haskell/app/Main.hs]: Fixed.

File: haskell/src/SegCFMM/TZIP16Metadata.hs (hunk 3 above)
Comment (rinn7e) [reply haskell/src/SegCFMM/TZIP16Metadata.hs]: Fixed.

File: README.md (hunk 4 above)
Comment (rinn7e) [reply README.md]: I've remove `known` and make other suggestion according to your feedback. This should be ready for re-review.

**Review by pasqu4le — APPROVED** [review]

**Pull request merged**